### PR TITLE
replace httpclient with net/http for evss letters and address

### PR DESCRIPTION
**Pull request by kreek**
- adds middleware that keeps custom http header keys immutable
- replaces httpclient with net/http as the faraday adapter for letters and pciu address services (other evss services still on custom base class not common client).
- removes unneeded memoization in controllers (requests always use a new instance of a controller)
- letters and address service base class initialized with user, user arg removed from methods

### CODE REVIEW

File: app/controllers/v0/addresses_controller.rb [hunk 1]
```diff
@@ -2,39 +2,39 @@
...
+      response = strategy.cache_or_service(:states) { service.get_states }
       render json: response,
              serializer: StatesSerializer
     end
 
     private
 
     def service
-      @service ||= EVSS::PCIUAddress::Service.new
+      EVSS::PCIUAddress::Service.new(@current_user)
```
Comment (jkassemi) [inline app/controllers/v0/addresses_controller.rb]: Minor yak shaving - not sure how the memoization here would have protected against anything as far as the request goes since, as you mentioned, a new instance of the controller is used for each request. If we'd left this then the service would have just been memoized for multiple calls within the same action. Prefer explicitly not memoizing it though.

File: lib/common/client/middleware/request/immutable_headers.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            Upgrade
+            User-Agent
+            Vary
+            Via
+          ).freeze
+
+          def call(env)
+            headers = {}
+            env.request_headers.each do |k, v|
+              if HTTP_PROTOCOL_HEADERS.include?(k)
```
Comment (jkassemi) [inline lib/common/client/middleware/request/immutable_headers.rb]: Minor, but lookup with include? is O(n) vs O(1) with a Set

File: lib/evss/configuration.rb [hunk 3]
```diff
@@ -46,8 +46,8 @@ def connection
         faraday.response :betamocks if mock_enabled?
         faraday.response :snakecase, symbolize: false
         faraday.response :json
-        faraday.use :remove_cookies
-        faraday.adapter :httpclient
+        faraday.use :immutable_headers
+        faraday.adapter Faraday.default_adapter
```
Comment (jkassemi) [inline lib/evss/configuration.rb]: :clap:

File: lib/evss/service.rb [hunk 4]
```diff
@@ -6,6 +6,15 @@ module EVSS
   class Service < Common::Client::Base
     STATSD_KEY_PREFIX = 'api.evss'
 
+    def initialize(user)
```
Comment (jkassemi) [inline lib/evss/service.rb]: I like how much cleanup this allows within the specs - definitely will remember to avoid caching the service object outside of a request cycle

File: lib/common/client/middleware/request/immutable_headers.rb (hunk 2 above)
Comment (jkassemi) [reply lib/common/client/middleware/request/immutable_headers.rb]: just ignore this - optimization here is seriously negligible 😉

File: spec/controllers/v0/letters_controller_spec.rb [hunk 5]
```diff
@@ -1,18 +0,0 @@
...
-
-  it 'should have a response that matches the schema' do
-    VCR.use_cassette('evss/letters/letters') do
-      request.headers['Authorization'] = "Token token=#{session.token}"
-      get :index
-      expect(response).to have_http_status(:ok)
-      expect(response).to match_response_schema('letters')
-    end
-  end
-end
```
Comment (lihanli) [inline spec/controllers/v0/letters_controller_spec.rb]: why was this deleted?

File: config/initializers/core_extensions.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# frozen_string_literal: true
+Dir[File.join(Rails.root, 'lib', 'core_extensions', '*.rb')].each { |file| require file }
```
Comment (lihanli) [inline config/initializers/core_extensions.rb]: how about putting immutable_string in lib or namespacing it with `module CoreExtensions` so that it gets autoloaded?

File: lib/common/client/middleware/request/immutable_headers.rb (hunk 2 above)
Comment (omgitsbillryan) [reply lib/common/client/middleware/request/immutable_headers.rb]: There's ~10'ish headers per request and this gets called on every request.  We might be talking picoseconds for all I know :laughing: , but it would give me a warm fuzzy to use a Set instead.

Comment (omgitsbillryan) [reply lib/common/client/middleware/request/immutable_headers.rb]: Also nitpicky, but using `.map!` instead of `.each` could save a few LoC:
```ruby
env.request_headers.map! do |k, v|
  if HTTP_PROTOCOL_HEADERS.exclude?(k)
    headers[ImmutableString.new(k)] = v
  end
end
```

File: spec/controllers/v0/letters_controller_spec.rb (hunk 5 above)
Comment (omgitsbillryan) [reply spec/controllers/v0/letters_controller_spec.rb]: :point_up:

File: lib/evss/configuration.rb (hunk 3 above)
Comment (omgitsbillryan) [reply lib/evss/configuration.rb]: A bit out of the loop - can you remind me the motivation here?

File: spec/controllers/v0/letters_controller_spec.rb (hunk 5 above)
Comment (kreek) [reply spec/controllers/v0/letters_controller_spec.rb]: @lihanli @omgitsbillryan removed because it's covered by https://github.com/department-of-veterans-affairs/vets-api/blob/master/spec/request/letters_request_spec.rb#L17-L26

File: lib/evss/configuration.rb (hunk 3 above)
Comment (kreek) [reply lib/evss/configuration.rb]: @omgitsbillryan the original reason for using HTTPClient was because Net::HTTP capitalizes headers and the VAAFI service does not follow [the spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) that headers should be case-insensitive and rejects them unless they're all lowercase 👎  

Subbing in Net::HTTP allows us to use the same client across all our services. Plus HTTPClient is a bit less performant and has an issue with its 🍪  manager 😧
